### ISSUE TITLE
ImagePreprocessor produces incorrect INT8 values (saturated at +127) (AIS-2173)

### ISSUE THREAD

**Issue by AlverGant** [espressif/esp-dl#267]
### Checklist

- [x] Checked the issue tracker for similar issues to ensure this is not a duplicate
- [x] Read the documentation to confirm the issue is not addressed there and your configuration is set correctly
- [x] Tested with the latest version to ensure the issue hasn't been fixed

### How often does this bug occurs?

always

### Expected behavior

The `dl::image::ImagePreprocessor` class is **not performing normalization correctly**. All input tensors are saturated at INT8 value `+127` regardless of actual image brightness, making the preprocessor completely unusable for models with ImageNet normalization.

## Expected Behavior

For a uniform grayscale image with **brightness=0.36** (RGB values ~92):

**Normalization formula**: `normalized = (pixel/255 - mean) / std`

```
RGB input: [0.36, 0.36, 0.36]
Normalized: [(0.36-0.485)/0.229, (0.36-0.456)/0.224, (0.36-0.406)/0.225]
          = [-0.546, -0.429, -0.204]
```

**Quantization to INT8** with scale=0.03125:
```
INT8 = round(normalized / scale)
     = round([-0.546/0.03125, -0.429/0.03125, -0.204/0.03125])
     = [-17, -14, -7]
```

**Expected statistics**:
- Center pixel: `[-17, -14, -7]`
- Min: -17, Max: -7
- Mean: ~-13

### Actual behavior (suspected bug)

**All input tensors show saturated values**:
```
Center pixel: [127, 127, 127]  ❌ Saturated at max INT8!
Min: -128, Max: 127
Mean: ~114 to ~122
Range: 255
```

**Values are CONSTANT regardless of brightness**:

| RAW Brightness | Expected INT8 | Actual INT8 | Status |
|----------------|---------------|-------------|--------|
| 36.7% | [-17, -14, -7] | [127, 127, 127] | ❌ WRONG |
| 40.7% | [-12, -9, -2] | [127, 127, 127] | ❌ WRONG |
| 71.0% | [+30, +35, +42] | [127, 127, 127] | ❌ WRONG |
| 84.5% | [+54, +60, +67] | [127, 127, 127] | ❌ WRONG |

### Error logs or terminal output

```shell

```

### Steps to reproduce the behavior

```cpp
#include "dl_image_preprocessor.hpp"
#include "dl_model_base.hpp"

// ImageNet normalization parameters
static const std::vector<float> MEAN_VALUES = {0.485f, 0.456f, 0.406f};
static const std::vector<float> STD_VALUES = {0.229f, 0.224f, 0.225f};

// Load model (quantized with ESP-PPQ, input scale=0.03125)
extern const uint8_t model_espdl[] asm("_binary_model_espdl_start");
dl::Model *model = new dl::Model((const char *)model_espdl,
                                  fbs::MODEL_LOCATION_IN_FLASH_RODATA);
model->build(0, dl::MEMORY_MANAGER_GREEDY);

// Create preprocessor with ImageNet normalization
dl::image::ImagePreprocessor *preprocessor =
    new dl::image::ImagePreprocessor(model, MEAN_VALUES, STD_VALUES);

// Preprocess image (128x128 RGB888)
dl::image::img_t img = {
    .data = camera_buffer,
    .width = 128,
    .height = 128,
    .pix_type = dl::image::DL_IMAGE_PIX_TYPE_RGB888
};

preprocessor->preprocess(img);

// Check input tensor - BUG: Always [127, 127, 127]!
dl::TensorBase *input_tensor = model->get_input();
int8_t *input_data = (int8_t *)input_tensor->data;
printf("Center pixel: [%d, %d, %d]\n",
       input_data[0], input_data[1], input_data[2]);
// Output: [127, 127, 127] instead of expected [-17, -14, -7]
```

## Debug Output

```
I exposure_analyzer: 📸 RAW IMAGE STATS: Brightness=0.367 (36.7%)
I exposure_analyzer: 🔍 INPUT TENSOR DEBUG:
I exposure_analyzer:   Exponent: -6    ← WRONG! Should be -5
I exposure_analyzer:   Center pixel INT8 (expected ~[-17,-14,-7]):
I exposure_analyzer:     RGB: [127, 127, 127]   ← SATURATED!
I exposure_analyzer:   INT8 Statistics:
I exposure_analyzer:     Min:  -128
I exposure_analyzer:     Max:  127      ← Should be ~-7 for bright=0.36
I exposure_analyzer:     Mean: 113.84   ← Should be ~-13
I exposure_analyzer:     Range: 255
```

## Impact

This bug makes `ImagePreprocessor` **completely unusable** for production models:

- ❌ Model always receives saturated input (+127)
- ❌ Model cannot distinguish different brightness levels
- ❌ Model output is constant (~0.85) regardless of input
- ❌ Cannot use ESP-DL for models requiring accurate preprocessing

### Project release version

latest master

### System architecture

Intel/AMD 64-bit (modern PC, older Mac)

### Operating system

Linux

### Operating system version

Linux Mint 22.1 Xia

### Shell

Bash

### Additional context

- **Quantization parameters** (ESP-PPQ) are correct: scale=0.03125, error=0.0079
- **PyTorch model** works perfectly with 100% accuracy on test images
- **ONNX conversion** is correct (verified output range 0.089-0.894)
- **Bug is ONLY in ESP-DL ImagePreprocessor**, not in model/quantization

**Comment by 100312dog** [comment]
@AlverGant Hi, here we use mean and std in range [0 - 255], not [0 - 1].
please try 
```
static const std::vector<float> MEAN_VALUES = {0.485f * 255, 0.456f * 255, 0.406f * 255};
static const std::vector<float> STD_VALUES = {0.229f * 255, 0.224f * 255, 0.225f * 255};
```
Not sure why ```Exponent: -6    ← WRONG! Should be -5.```
The exponent of imagepreprocessor comes from the model. And the model exponent comes from ppq, they should be same.

https://docs.espressif.com/projects/esp-dl/en/latest/tutorials/how_to_quantize_model.html#quantize-and-export-espdl
Check .info file to see the exponent after ppq quantization.

**Comment by AlverGant** [comment]
@100312dog Hi,

Thank you for the quick response and guidance! I can confirm that **your solution works perfectly**. Here are the results:

  ### 1. Fixed Normalization Range ✅

  Changed from `[0-1]` to `[0-255]` range as you suggested:

  ```cpp
  // Before (WRONG):
  static const std::vector<float> MEAN_VALUES = {0.485f, 0.456f, 0.406f};
  static const std::vector<float> STD_VALUES = {0.229f, 0.224f, 0.225f};

  // After (CORRECT):
  static const std::vector<float> MEAN_VALUES = {0.485f * 255, 0.456f * 255, 0.406f * 255};  // [123.675, 116.280, 103.530]
  static const std::vector<float> STD_VALUES = {0.229f * 255, 0.224f * 255, 0.225f * 255};   // [58.395, 57.120, 57.375]

  2. Validation Results ✅

  Input tensor is now correctly normalized:
  - Values range from -128 to +127 (NOT saturated at 127!)
  - Example: Center pixel = [-17, -14, -7] for brightness=0.36 ✓
  - Mean values vary dynamically: -22.43, 4.66, 14.71, etc. ✓

  Model output is now dynamic:
  - Brightness varies: 0.428 → 0.577 (NOT constant!)
  - System successfully adjusts camera exposure/gain ✓

  3. Exponent Verification ✅

  You were absolutely right about the exponent issue!

  Initial problem: Our model had output exponent = -6 (incorrect)

  Root cause: Model was poorly quantized with random weights

  Solution:
  - Re-trained model from scratch (7192 images, 20 epochs, val_loss=0.0025)
  - Re-quantized with ESP-PPQ using 1100 calibration images
  - Result: Output exponent is now -5 ✓

  Verified in .info file:
  %input[INT8, 1x128x128x3], exponents: [-6]
  %output[INT8, 1x3], exponents: [-5]  ✅ CORRECT!
